text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,475 @@
+app.get("/my-route", function () {
+  const isolationScope = Sentry.getIsolationScope();
+  isolationScope !== initialIsolationScope;
+  // Isolation scope inherits from previously current isolation scope
+  isolationScope.getTags(); // -> tag1: aa
+});
+```
+
+Note that in environments where you do not have multiple execution contexts (e.g. Browser JS), there is basically only a single isolation scope that is fundamentally the same as the global scope. In these environments, global & isolation scope can be used completely interchangeably. However, for consistency we still have these concepts everywhere - but users do not need to care about them, and we can reflect this in the SDK specific documentation.
+
+The isolation scope always inherits from the previously active isolation scope.
+
+### How does the isolation scope differ from the current scope?
+
+If you are unsure why this exists, here an example:
+
+```js
+app.get('/my-route', function() {
+  Sentry.withScope(scope1 => {
+
+  })
+})
+```
+
+### How to document isolation scopes
+
+Isolation scopes will be documented in a new section under "Enriching Events > Per request data" (or similar, based on SDK language & needs), which will explain how to use `getIsolationScope()` and when/how data added there will be applied.
+
+Global methods to set data, like `Sentry.setTag()` or `Sentry.setUser()`, will set on the isolation scope. Only if you want to set on a different scope, you need to manually get it (e.g. `getGlobalScope().setTag()`).
+
+### When to create an isolation scope
+
+For most server-side SDKs, an isolation scope will be created for each request being processed.
+Roughly, it will equate to each time we currently fork a hub.
+
+### Examples for isolation scopes
+
+Example for instrumentation that we would write:
+
+```ts
+function wrapHttpServerRequest(original: Function): Function {
+  // Fork an execution context for this server request, that is isolated
+  return Sentry.withIsolatedScope((scope) => {
+    // anything in here will have the same isolated scope!
+    return original();
+  });
+}
+```
+
+Example for hooking into external auto-instrumentation (e.g. OpenTelemetry):
+
+```ts
+let onRequestHook: (span: Span) => void;
+
+// This method is not defined by us, but is some external code
+// Here just for demonstration purposes of how that may be implemented
+function otelWrapHttpServerRequest(original: Function): Function {
+  // Fork an execution context for this server request,
+  // but without isolating this!
+  return Sentry.withScope((scope) => {
+    onRequestHook(trace.getActiveSpan());
+    return original();
+  });
+}
+
+// This would be our custom sentry configuration
+onRequestHook = () => {
+  const scope = getCurrentScope();
+  scope.isolate(); // Add an isolation scope to the already forked scope
+};
+```
+
+## Applying scopes
+
+Scopes are applied in this order to events:
```

Comment:
Given the example on how scopes are merged I'm wondering what happens the other way round when scopes are read? E.g. Would `Sentry.getCurrentScope().getRelease()` return a value when the release was set on the global scope ~ `Sentry.getGlobalScope().setRelease("1.0.0")`

Reply:
It's a good question! I would say it should only return the value on the actual scope being accessed. If you want to have some cascading behavior it should be a top-level method like `Sentry.getRelease()` or something like this?